win32: Update dosinst.c

k-takata opened the pull request:
As I mentioned in #3485 and #3501, there are some issues in dosinst.c.
@brammool I don't think you are going to merge #3501 soon, but how about only merging dosinst.c part soon?
This PR extracts only dosinst.c changes from #3501 (by using `git cherry-pick`).

This PR includes 5 commits. (If you really needed, you can merge only some of them.)

1. Fixes #3485. Register DisplayIcon, DisplayVersion and Publisher for the uninstaller's information.
  https://github.com/vim/vim/commit/77bce1333d0b9ea0520c45fdb6ef316ea1d0feaa.patch
2. Don't set `'diffexpr'` in the default `_vimrc` if the internal diff is supported.
  https://github.com/vim/vim/commit/e69597ae1351b9670905a8e33724aca0aca3ada3.patch
3. Support setting Vi compatibility in the default `_vimrc` from command line.
  https://github.com/vim/vim/commit/2348655cfec4c9f4779ced0f38e3dddc6a9164c7.patch
4. Remove needless sleeps.
  `EnumWindows()` is not an asynchronous function. No need to sleep after the function.
  No need to wait 3 seconds after showing "Done!" when the stdout is not a tty.
  https://github.com/vim/vim/commit/fb743bed597c13d77dd319ae912bdb25a6c5b4d0.patch
5. Add some comments to the default `_vimrc`.
  https://github.com/vim/vim/commit/9ccf759a4157236481c096fa73e9583224f49ed9.patch

(Note: No. 5 depends on No. 2.)

k-takata committed:
dosinst: Register some more information in installer

Register the following items:
* DisplayIcon
* DisplayVersion
* Publisher

EstimatedSize will be registered in the NSIS script, because it is
easier than registering in dosinst.c.

k-takata committed:
dosinst: Adjust the default _vimrc

Don't set 'diffexpr' if the internal diff is supported.

k-takata committed:
dosinst: Support setting Vi compatibility from command line

Also add nocp setting.

k-takata committed:
dosinst: Remove needless sleeps

EnumWindows() is not an asynchronous function. No need to sleep after
the function.

No need to wait 3 seconds after showing "Done!" when the stdout is not
a tty.

k-takata committed:
dosinst: Add some comments to the default _vimrc

brammool commented:
Ken Takata wrote:

> As I mentioned in #3485 and #3501, there are some issues in dosinst.c.
> @brammool I don't think you are going to merge #3501 soon, but how about only merging dosinst.c part soon?
> This PR extracts only dosinst.c changes from #3501 (by using `git cherry-pick`).
> 
> This PR includes 5 commits. (If you really needed, you can merge only some of them.)
> 
> 1. Fixes #3485. Register DisplayIcon, DisplayVersion and Publisher for the uninstaller's information.
>   https://github.com/vim/vim/commit/77bce1333d0b9ea0520c45fdb6ef316ea1d0feaa.patch
> 2. Don't set `'diffexpr'` in the default `_vimrc` if the internal diff is supported.
>   https://github.com/vim/vim/commit/e69597ae1351b9670905a8e33724aca0aca3ada3.patch
> 3. Support setting Vi compatibility in the default `_vimrc` from command line.
>   https://github.com/vim/vim/commit/2348655cfec4c9f4779ced0f38e3dddc6a9164c7.patch
> 4. Remove needless sleeps.
>   `EnumWindows()` is not an asynchronous function. No need to sleep after the function.
>   No need to wait 3 seconds after showing "Done!" when the stdout is not a tty.
>   https://github.com/vim/vim/commit/fb743bed597c13d77dd319ae912bdb25a6c5b4d0.patch
> 5. Add some comments to the default `_vimrc`.
>   https://github.com/vim/vim/commit/9ccf759a4157236481c096fa73e9583224f49ed9.patch
> 
> (Note: No. 5 depends on No. 2.)

Thanks.  We should indeed break it up in smaller parts, if that makes
sense.  I'll look into this PR.

-- 
Wizards had always known that the act of observation changed the thing that
was observed, and sometimes forgot that it also changed the observer too.
			Terry Pratchett  -  Interesting times

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

k-takata commented:
Sorry, I updated the 2nd and the 5th commits, because the 2nd commit mistakenly removed "\n".
The total changes by this PR is not changed by this update.
Updated commits are as follows:

1. Fixes #3485. Register DisplayIcon, DisplayVersion and Publisher for the uninstaller's information.
  https://github.com/vim/vim/commit/77bce1333d0b9ea0520c45fdb6ef316ea1d0feaa.patch
2. Don't set `'diffexpr'` in the default `_vimrc` if the internal diff is supported.
  https://github.com/vim/vim/commit/f0de3698a34e282080babadfeaae016c789252aa.patch
3. Support setting Vi compatibility in the default `_vimrc` from command line.
  https://github.com/vim/vim/commit/5f0390c20da2d7108d3fb535e4762b14587727ce.patch
4. Remove needless sleeps.
  `EnumWindows()` is not an asynchronous function. No need to sleep after the function.
  No need to wait 3 seconds after showing "Done!" when the stdout is not a tty.
  https://github.com/vim/vim/commit/cbc819b399ff4a824b3dac5cea9c7533332e4356.patch
5. Add some comments to the default `_vimrc`.
  https://github.com/vim/vim/commit/5b39b426acbd3d39fab684f2621329478f6e8eae.patch

(Note: No. 5 depends on No. 2 and also No. 3.)
